lib/AST/ArchetypeBuilder.cpp

```diff
@@ -467,13 +468,46 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
   // archetype conforms.
   for (auto &conforms : ConformsTo) {
     for (auto member : conforms.first->lookupDirect(nestedName)) {
-      auto assocType = dyn_cast<AssociatedTypeDecl>(member);
-      if (!assocType)
+      PotentialArchetype *pa;
+      
+      if (auto assocType = dyn_cast<AssociatedTypeDecl>(member)) {
+        // Resolve this nested type to this associated type.
+        pa = new PotentialArchetype(this, assocType);
+      } else if (auto alias = dyn_cast<TypeAliasDecl>(member)) {
+        // Resolve this nested type to this type alias.
+        pa = new PotentialArchetype(this, alias);
+        
+        if (alias->hasUnderlyingType()) {
+          auto type = alias->getUnderlyingType();
+          if (type->hasArchetype()) {
+            auto repr = alias->getUnderlyingTypeLoc().getTypeRepr();
+            
+            if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+              PotentialArchetype *existingPA = this;
+              for (auto component : identRepr->getComponentRange()) {
+                existingPA = existingPA->getNestedType(component->getIdentifier(),
```

Comment:
@DougGregor I can see how to do it the archetype decomposing way, and I can write that code. I'm not sure how interface types work, and it looks like computeInterfaceType() is on NominalDecl, which a TypeAliasDecl isn't. The fallout from changing TypeAliasDecl's superclass is beyond what I'm currently comfortable with. Sound okay to do archetype decomposing, and save interface type for future enhancement?